File: tv/apple-tv/PlayBridge TV/PlayBridge TV/PlaybackEngine.swift
import Combine
import Foundation
import PlayBridgeProtocol

/// Playback states that are common across all engines.
public enum PlaybackState: String {
    case idle
    case loading
    case playing
    case paused
    case buffering
    case stopped
    case ended
    case error
}

/// Common interface for video playback engines (AVPlayer, VLC, etc.)
protocol PlaybackEngine: AnyObject {
    /// Publisher for current playback state updates
    var state: AnyPublisher<PlaybackState, Never> { get }

    /// Publisher for current position (seconds) updates
    var position: AnyPublisher<TimeInterval, Never> { get }

    /// Total duration of the current media (seconds)
    var duration: TimeInterval { get }

    /// Load a new media payload
    func load(_ payload: PlayPayload) async throws

    /// Start playback
    func play()

    /// Pause playback
    func pause()

    /// Stop playback and release resources
    func stop()

    /// Seek to a specific position (seconds)
    func seek(to: TimeInterval) async

    /// Change playback rate
    func setRate(_ rate: Float)

    /// Change audio track
    func setAudioTrack(_ id: String?)

    /// Change subtitle track
    func setSubtitleTrack(_ id: String?)

    /// Attach an external subtitle file
    func attachExternalSubtitle(url: URL) async throws

    /// Set color filters
    func setFilter(_ settings: ColorFilterSettings)
}

File: shared/src/commonMain/kotlin/com/playbridge/shared/player/PlaybackEngine.kt
package com.playbridge.shared.player

import playbridge.PlayPayload
import kotlinx.coroutines.flow.StateFlow

/**
 * Common interface for video playback engines across Android and Apple TV.
 */
interface PlaybackEngine {
    val state: StateFlow<PlaybackState>
    val position: StateFlow<Long>     // ms
    val duration: StateFlow<Long>     // ms, -1 if unknown
    val audioTracks: StateFlow<List<Track>>
    val subtitleTracks: StateFlow<List<Track>>
    var isTransitioning: Boolean

    suspend fun load(payload: PlayPayload)
    fun play()
    fun pause()
    fun stop()
    fun seek(positionMs: Long)
    fun setRate(rate: Float)
    fun setAudioTrack(id: String?)
    fun setSubtitleTrack(id: String?)
    suspend fun attachExternalSubtitle(url: String, language: String?)
    fun setFilter(filter: VideoFilter, customParams: List<Float>? = null)
    fun release()
}

sealed class PlaybackState {
    data object Idle         : PlaybackState()
    data object Buffering    : PlaybackState()
    data object Ready        : PlaybackState()
    data object Playing      : PlaybackState()
    data object Paused       : PlaybackState()
    data object Ended        : PlaybackState()
    data class  Error(val code: String, val msg: String) : PlaybackState()
}

data class Track(val id: String, val label: String, val language: String?)

enum class VideoFilter(val label: String) {
    NONE("None"),
    HDR("HDR"),
    NIGHT("Night"),
    MOVIE("Movie"),
    CINEMA("Cinema"),
    ACTION("Action"),
    DEEP_BLACK("Deep Black"),
    GRAYSCALE("Grayscale"),
    VIVID("Vivid"),
    CUSTOM("Custom")
}
